---
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(ggplot2)
library(gridExtra)
library(ggpubr)
```

```{r}
# Setting initial parameters
df_ab <- read.table("/var/home/crestang/files/work/genomic_analysis/data/snp_tables/filtered_snps_ab.txt", header = TRUE)
df_abc <- read.table("/var/home/crestang/files/work/genomic_analysis/data/snp_tables/filtered_snps_abc.txt", header = TRUE)
df_abcd <- read.table("/var/home/crestang/files/work/genomic_analysis/data/snp_tables/filtered_snps_abcd.txt", header = TRUE)
```

```{r}
# Replaces "." with 0 and converts columns 6 to the end to numeric

fix_dots <- function(mydf) {
  mydf[mydf == "."] <- 0
  mydf[, 6:ncol(mydf)] <- lapply(mydf[, 6:ncol(mydf)], as.numeric)
  return(mydf)
}

df_ab <- fix_dots(df_ab)
df_abc <- fix_dots(df_abc)
df_abcd <- fix_dots(df_abcd)

# Functions
avg_cov_per_sample <- function() {
  all_cov <- df[grep("N_", colnames(df))]
  mean_cov <- colMeans(all_cov)
  names(mean_cov) <- gsub("N_", "", gsub("_rep0", "", gsub("_gen", "_G", names(mean_cov))))
  names(mean_cov) <- gsub("G01", "G1", names(mean_cov))
  #print(mean_cov)
  return(mean_cov)
}

min_cov_per_sample <- function() {
  all_cov <- df[grep("N_", colnames(df))]
  min_cov <- apply(all_cov, 2, min)
  names(min_cov) <- gsub("N_", "", gsub("_rep0", "", gsub("_gen", "_G", names(min_cov))))
  names(min_cov) <- gsub("G01", "G1", names(min_cov))
  #print(min_cov)
  return(min_cov)
}

max_cov_per_sample <- function() {
  all_cov <- df[grep("N_", colnames(df))]
  max_cov <- apply(all_cov, 2, max)
  names(max_cov) <- gsub("N_", "", gsub("_rep0", "", gsub("_gen", "_G", names(max_cov))))
  names(max_cov) <- gsub("G01", "G1", names(max_cov))
  #print(max_cov)
  return(max_cov)
}

# Initializing report df
generate_report_df <- function() {
  mean_cov <- avg_cov_per_sample()
  cov_df <- enframe(mean_cov)
  colnames(cov_df) <- c("sample", "Average_Coverage")
  #cov_df <- mean_cov_df[order(mean_cov_df$No_Filter_Average_Coverage),]
  cov_df$Min_Coverage <- min_cov_per_sample()
  cov_df$Max_Coverage <- max_cov_per_sample()
  #sort by average coverage
  cov_df <- cov_df[order(cov_df$Average_Coverage),]

  return(cov_df)
}

df <- df_abcd
```

## Gio's Project - Coverage Report (2024-07-26)
## Raw SNP table
```{r}
# No filtering

cov_df <- generate_report_df()

# Print knitr table and number of SNPs
cat(paste("Initial number of reads:" , format(nrow(df), big.mark = ",")))

knitr::kable(cov_df, format = "markdown", caption = "Coverage per Sample - No filtering")
```
\pagebreak  

## Keep only chromosomes 2L, 2R, 3L, 3R, X
```{r}
# Filter chromosomes
nrow_before <- nrow(df)

chromosomes = c("2L", "2R", "3L", "3R", "X")
df <- df[df$CHROM %in% chromosomes,]

cov_df <- generate_report_df()


nrow_after <- nrow(df)

dif <- nrow_before - nrow_after


# Print knitr table and number of SNPs
print(paste("Number of reads before filtering:" , format(nrow_before, big.mark = ",")))
print(paste("Number of reads after filtering:" , format(nrow_after, big.mark = ",")))
print(paste("SNPs removed:" , format(dif, big.mark = ",")))
knitr::kable(cov_df, format = "markdown", caption = "Coverage per Sample - Chromosome filtering")
```
\pagebreak

## Mininimum genome-wide coverage of 30
```{r}
# Apply a min filter of 30
nrow_before <- nrow(df)

min_cov <- 30
df <- df %>% filter_at(vars(starts_with("N_")), all_vars(. >= min_cov))

cov_df <- generate_report_df()

nrow_after <- nrow(df)

dif <- nrow_before - nrow_after

# Print knitr table and number of SNPs
print(paste("Number of reads before filtering:" , format(nrow_before, big.mark = ",")))
print(paste("Number of reads after filtering:" , format(nrow_after, big.mark = ",")))
print(paste("SNPs removed:" , format(dif, big.mark = ",")))
knitr::kable(cov_df, format = "markdown", caption = "Coverage per Sample - Min coverage filtering")
```
\pagebreak

## Maximum genome-wide coverage of 300
```{r}
# Apply a max filter of 300
nrow_before <- nrow(df)

max_cov <- 500
df <- df %>% filter_at(vars(starts_with("N_")), all_vars(. <= max_cov))

cov_df <- generate_report_df()

nrow_after <- nrow(df)

dif <- nrow_before - nrow_after


# Print knitr table and number of SNPs
print(paste("Number of reads before filtering:" , format(nrow_before, big.mark = ",")))
print(paste("Number of reads after filtering:" , format(nrow_after, big.mark = ",")))
print(paste("SNPs removed:" , format(dif, big.mark = ",")))
knitr::kable(cov_df, format = "markdown", caption = "Coverage per Sample - Max coverage filtering")
```
\pagebreak

## MAF of 0.025
```{r}
# Apply a MAF of 0.025
nrow_before <- nrow(df)

samples <- unique(gsub("alt_|N_", "", colnames(df[grep("alt_|N_", colnames(df),)])))
df$freq_rowsums <- rowSums(df[paste("alt_", samples, sep="")])/rowSums(df[paste("N_", samples, sep="")])
df <- df[df$freq_rowsums >= 0.025 & df$freq_rowsums <= 0.975,]

cov_df <- generate_report_df()

nrow_after <- nrow(df)

dif <- nrow_before - nrow_after

print(paste("Number of reads before filtering:" , format(nrow_before, big.mark = ",")))
print(paste("Number of reads after filtering:" , format(nrow_after, big.mark = ",")))
print(paste("SNPs removed:" , format(dif, big.mark = ",")))
knitr::kable(cov_df, format = "markdown", caption = "Coverage per Sample - MAF cutoff of 0.025")

```